### DISCUSSION TITLE
[Question] Copy Highlights and Annotations from one pdf to another

### DISCUSSION THREAD

**Discussion by Mark-Joy** [pymupdf/PyMuPDF#1561] [Q&A]
Is there an easy way to copy Highlights and Annotations from one pdf to another?
Could you please lay out the necessary things that have to be taken care of?

**Comment by JorjMcKie** [comment]
Hard to imagine how this may make sense at all: annotations are page-specific after all.
So, if the two PDFs are unrelated, how would you decide that the annots of page x of PDF 1 should go to page y of PDF 2?

But even assuming we had an algorithm for that, general annotations can encompass all sorts of things, so that I would argue it is not possible at all, or require a **_very_** significant effort.
Highlights are a fairly simple case of annotations, so maybe there is a doable way. But again: what sense would that make, if text on the input page is being highlighted, where there may be no text at all on the output page?

**Reply by Mark-Joy** [reply]
Here is an use case: 
Suppose I have an `original pdf`.
I made some changes to it that can't be reverted, bloated its size. We can call this `modified pdf`. Also, I made some highlights and annotation during my reading.
Now I want to copy those highlights and annotations back to the `original pdf`.

**Comment by JorjMcKie** [comment]
Ok, I understand better now.
Any details on those unwanted / bloating changes:
Are they affecting the same pages where you created annotations? If not, you could copy single pages of modified.pdf back and replace the respective original pages.

Otherwise you must iterate thru the annotations of modified.pdf's pages, read out its definitions and re-insert each in the corresponding original page.

**Reply by Mark-Joy** [reply]
Thank you for your suggestion.
I will try to write a python script this weekend.
Will also upload it here.

**Comment by Mark-Joy** [comment]
@JorjMcKie,
I see Page class has many methods to add an annot such as `add_highlight_annot`, `add_freetext_annot`. Obviously, to use those methods, the annot's type must be known.
Is there a way to add a general annotation like:
`ori_pdf[p_cnt].add_annot(annot)`

```python
p_cnt=0
for page in modified_pdf:
    for annot in page.annots():
        ori_pdf[p_cnt].add_annot(annot)
    p_cnt+=1
```

**Reply by JorjMcKie** [reply]
No, there is not. I also see no major problem here, because for every annot yielded you can determine its type - a unique integer.
Make a mapping of that type to the corresponding page add_*_annot() method ...

**Reply by JorjMcKie** [reply]
you probably also know which types of annots you have been adding at all. Probably not all types do occur here.
If you file has annots from others origins than PyMuPDF, please be aware that PyMuPDF is not capable to exactly replicate them by its own methods, e.g. a FreeText annot made by PyMuPDF will never look like one created by another tool ...
